src/CodeGen_X86.cpp

```diff
@@ -480,14 +486,16 @@ void CodeGen_X86::codegen_vector_reduce(const VectorReduce *op, const Expr &init
         Expr pattern;
         const char *intrin;
         Type narrow_type;
+        bool combine_init;
     };
     // clang-format off
     static const Pattern patterns[] = {
-        {2, i32(widening_mul(wild_i16x_, wild_i16x_)), "pmaddwd", Int(16)},
-        {2, i32(widening_mul(wild_i8x_, wild_i8x_)), "pmaddwd", Int(16)},
-        {2, i32(widening_mul(wild_i8x_, wild_u8x_)), "pmaddwd", Int(16)},
-        {2, i32(widening_mul(wild_u8x_, wild_i8x_)), "pmaddwd", Int(16)},
-        {2, i32(widening_mul(wild_u8x_, wild_u8x_)), "pmaddwd", Int(16)},
+        {2, f32(wild_bf16x_) * f32(wild_bf16x_), "dot_product", BFloat(16), true},
```

Comment:
I think this should just be `wild_f32x_ * wild_f32x_`. The operands will be `lossless_cast`ed to BFloat(16), which looks like it should work (https://github.com/halide/Halide/blob/master/src/Type.cpp#L118).